README.md

PCString
=========
maintained by [PandaClouds.com](https://pandaclouds.com)

`PCString` is a lightweight JavaScript library for Node.js that provides additional String methods.


Installation
------------

1. If you want to use this library, you first need to install the [Node.js](https://nodejs.org/en/).

2. When you install node.js, will also be installed [npm](https://www.npmjs.com/).

3. Please run the following command.

```
npm install --save @panda-clouds/string
```

Usage
-----

### Node.js

```javascript
const PCString = require('@panda-clouds/string');

// example usage
PCString.isString('yup!'); // => true;
PCString.hasWhitespace('ABC'); // => false;
```

You can replace PCString with any variable.


Methods
-------

[Unit Tests] are an additional resource for learning functionality.

### - isString(string)

Returns whether a given object is a String.

Example:

```javascript
PCString.isString('ABC') // => true
PCString.isString(5) // => false
PCString.isString({}) // => false
PCString.isString([]) // => false
```

### - hasWhiteSpace(string)

returns true if the string has white space, false if not.


Example:

```javascript
PCString.hasWhitespace(' ') // => true
PCString.hasWhitespace('A B') // => true
PCString.hasWhitespace('AB') // => false
PCString.hasWhitespace('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-!@#$%^&*()') // => false
```


### - removeWhitespace(string)

Removes all white characters from the given string.

Example:

```javascript
PCString.removeWhitespace(' A B  C   D    ') // => 'ABCD'
PCString.removeWhitespace('	A	B	C			D	') // => 'ABCD'
PCString.removeWhitespace('\nA\nB\n\nC\n\n\nD\n\n\n\n') // => 'ABCD'
PCString.removeWhitespace(null) // => ''
```

### - domainSafeString(string)

Removes all non-domain-safe characters.

Example:

```javascript
PCString.domainSafeString('_A_B__C___D____') // => 'ABCD'
PCString.domainSafeString('	A	B	C			D	') // => 'ABCD'
PCString.domainSafeString('\nA\nB\n\nC\n\n\nD\n\n\n\n') // => 'ABCD'
PCString.domainSafeString(null) // => ''
```


Contributions
-------------

Pull requests are welcome! here is a checklist to speed things up:

- modify `PCString.js`
- add unit tests in `PCString.spec.js`
- run `npm test`
- document method in `README.md`
- add your name to 'Contributors' in `README.md`


### Contributors

(Add your name)

- [*] [Marc Smith](https://github.com/mrmarcsmith)


[Unit Tests]: https://github.com/panda-clouds/string/blob/master/spec/PCString.spec.js


CHANGELOG.md
## PCAddressFormatter Changelog

### master
[Full Changelog](https://github.com/panda-clouds/address-formatter/compare/2.0.0...master)

### 0.19.0

- moved from parse-jasmine to parse-runner
- linted to conform to Panda Clouds config
- moved from jasmine to jest

### 0.18.0

- added zipcode and city to spoof
- bumped geocoder to 1.10.0

### 0.17.0

- added zipcode and city to spoof
- bumped geocoder to 1.9.0

### 0.16.1

- bumped formatter to 2.4.2
- bumped geocoder to 1.8.1

### 0.16.0

- bumped geocoder to 1.8.0
- bumped formatter to 2.4.1

### 0.15.0

- fixed bug where creator wouldn't be saved in preview
- bumped geocoder to 1.7
- bumped foramtter to 2.4.0
- added 'notes' property
- added 'gateCode' property
- added 'unit' property
- added 'plusFour' property

### 0.14.0

- bumped geocoder to 1.6.0

### 0.13.0

- bumped formatter to 2.3.0

### 0.12.0

- bumped formatter to 2.2.1
- removed logs

### 0.11.0

- added useMasterKey and limit query to "user"

### 0.10.6

- added runtime switching between prod and testing parse

### 0.10.5

- added Parse Jasmine unit tests

### 0.10.4

- testing NODE_ENV

### 0.10.3

- removed var Parse = ... in PCAddressPreview

### 0.10.2

- moved parse dependency to "dev"

### 0.10.1

- fixed "this." in static function

### 0.9.0

- set radius default to 1 mile

### 0.9.0

- added logs

### 0.8.0

- bumped geocoder to 1.4.0

### 0.7.0

- bumped geocoder to 1.3.0

### 0.6.0

- added logs

### 0.2.0

- added the "street" property of PCGeocoder

spec/PCAddressPreview.spec.js
const PCAddressPreview = require('../src/PCAddressPreview.js');
const PCParseJasmine = require('@panda-clouds/parse-runner');
const cloud =
`
Parse.Cloud.define('challenge', function(request, response) {
  response.success('everest');
});
`;

describe('test PCAddress.js', () => {
	const parseRunner = new PCParseJasmine();

	parseRunner.cloud(cloud);

	beforeAll(async () => {
		await parseRunner.startParseServer();
	}, 1000 * 60 * 2);

	afterAll(async () => {
		await parseRunner.cleanUp();
	});

	describe('_generateRandomFloat', () => {
		/* eslint-disable jest/prefer-expect-assertions */
		it('should hit 0,1 and 2 in 100 interations', () => {
			// expect.assertions(303); doesn't work because we stop after finding the number
			let found0 = false;
			let found1 = false;
			let found2 = false;
			const iterations = 100;

			for (let i = 0; i < iterations; i++) {
				const result = PCAddressPreview._generateRandomFloat(0, 2, 6);

				expect(result).not.toBeLessThan(0);
				expect(result).not.toBeGreaterThan(2);
				expect(result.lenth).toBe();

				if (Math.round(result) === 0) {
					found0 = true;
				} else if (Math.round(result) === 1) {
					found1 = true;
				} else if (Math.round(result) === 2) {
					found2 = true;
				}

				if (found0 && found1 && found2) {
					break;
				}
			}

			expect(found0).toBe(true);
			expect(found1).toBe(true);
			expect(found2).toBe(true);


			// Special matcher from
			// https://github.com/JamieMason/Jasmine-Matchers
		});

		it('should hit -2,-1,0,1 and 2 in 100 interations', () => {
			// expect.assertions(206); doesn't work because we stop after finding the number
			let foundNeg2 = false;
			let foundNeg1 = false;
			let found0 = false;
			let found1 = false;
			let found2 = false;
			const iterations = 100;

			for (let i = 0; i < iterations; i++) {
				const result = PCAddressPreview._generateRandomFloat(-2, 2, 6);

				expect(result).not.toBeLessThan(-2);
				expect(result).not.toBeGreaterThan(2);

				if (Math.round(result) === -2) {
					foundNeg2 = true;
				} else if (Math.round(result) === -1) {
					foundNeg1 = true;
				} else if (Math.round(result) === 0) {
					found0 = true;
				} else if (Math.round(result) === 1) {
					found1 = true;
				} else if (Math.round(result) === 2) {
					found2 = true;
				}

				if (foundNeg2 && foundNeg1 && found0 && found1 && found2) {
					break;
				}
			}
			expect(foundNeg2).toBe(true);
			expect(foundNeg1).toBe(true);
			expect(found0).toBe(true);
			expect(found1).toBe(true);
			expect(found2).toBe(true);

			// Special matcher from
			// https://github.com/JamieMason/Jasmine-Matchers
		});
		/* eslint-enable jest/prefer-expect-assertions */
	});


	describe('_randomPointWithInRadiusInMiles', () => {
		it('should randomize 1341 W 10th Place with no radius', () => {
			expect.assertions(1000);
			for (let x = 0; x < 500; x++) {
				const preview = new PCAddressPreview();

				preview.street('1341 W 10th Place');
				preview.city('Tempe');
				preview.state('AZ');
				preview.country('US');
				preview.zipcode('85281');

				// Bamboozle up a Geo-Point
				const knownHouseGeo = PCAddressPreview.GeoPoint(33.417847, -111.960097);
				const spoofGeo = PCAddressPreview._randomPointWithInRadiusInMiles(knownHouseGeo);

				const spoofDelta = knownHouseGeo.milesTo(spoofGeo);

				// the spoof geo should be within 1 mile of our house
				expect(spoofDelta).not.toBeLessThan(0.02);
				expect(spoofDelta).not.toBeGreaterThan(1);
			}
		});

		it('should randomize 1341 W 10th Place with 2 mile radius', () => {
			expect.assertions(1000);
			for (let x = 0; x < 500; x++) {
				const preview = new PCAddressPreview();

				preview.street('1341 W 10th Place');
				preview.city('Tempe');
				preview.state('AZ');
				preview.country('US');
				preview.zipcode('85281');

				// Bamboozle up a Geo-Point
				const knownHouseGeo = PCAddressPreview.GeoPoint(33.417847, -111.960097);
				const spoofGeo = PCAddressPreview._randomPointWithInRadiusInMiles(knownHouseGeo, 100);

				const spoofDelta = knownHouseGeo.milesTo(spoofGeo);

				// the spoof geo should be within 1 mile of our house
				expect(spoofDelta).not.toBeLessThan(0.02);
				expect(spoofDelta).not.toBeGreaterThan(100);
			}
		});
	});


	describe('preview', () => {
		it('should geocode an 1341', async () => {
			expect.assertions(21);
			const street1 = '1341 W 10th Pl';
			const city1 = 'Tempe';
			const state1 = 'Arizona';
			const country1 = 'USA';
			const zipcode1 = '85281';

			// Attempt to trick the system into thinking
			// that it is a different address
			const street2 = '1341 west 10th Place';
			const city2 = 'teMpe';
			const state2 = 'aZ';
			const country2 = 'United states of amEriCa';
			const zipcode2 = '85281';

			const knownLat = 33.417847;
			const knownLong = -111.959508;
			const knownHouseGeo = PCAddressPreview.GeoPoint(knownLat, knownLong);
			const closeEnoughLat = 33.417;
			const closeEnoughLong = -111.959;

			const user = {
				__type: 'Pointer',
				className: '_User',
				objectId: 'abc123',
			};

			// check that caches results are the same
			const address = new PCAddressPreview();

			address.street(street1);
			address.city(city1);
			address.state(state1);
			address.country(country1);
			address.zipcode(zipcode1);
			address.user(user);
			// address.previewRadius(1); default

			const result = await address.savePreview();

			// eslint-disable-next-line no-console
			console.log('ncioed' + JSON.stringify(result));
			expect(result).toBeDefined();
			expect(result.className).toBe('PCAddressPreview');
			expect(result.id).toHaveLength(10);
			const spoofId = result.id;
			const spoof = result.get('geoPoint');

			expect(spoof).toBeDefined();

			const pcAddress = result.get('address');

			expect(pcAddress).toBeDefined();

			const actualGeo = pcAddress.get('geoPoint');

			expect(actualGeo).toBeDefined();


			expect(actualGeo.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta = knownHouseGeo.milesTo(actualGeo);

			expect(actualGeoDelta).toBeLessThan(0.1);

			const spoofDelta = knownHouseGeo.milesTo(spoof);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta).toBeLessThan(1.1);

			// check that caches results are the same
			const address2 = new PCAddressPreview();

			address2.street(street2);
			address2.city(city2);
			address2.state(state2);
			address2.country(country2);
			address2.zipcode(zipcode2);
			address2.user(user);
			// address2.previewRadius(1); default

			const result2 = await address2.savePreview();

			// eslint-disable-next-line no-console
			console.log('dif' + JSON.stringify(result2));
			expect(result2).toBeDefined();
			expect(result2.className).toBe('PCAddressPreview');
			expect(result2.id).toHaveLength(10);
			// ensures the cache was used
			expect(result2.id).toBe(spoofId);
			const spoof2 = result2.get('geoPoint');

			expect(spoof2).toBeDefined();

			const pcAddress2 = result2.get('address');

			expect(pcAddress2).toBeDefined();

			const actualGeo2 = pcAddress2.get('geoPoint');

			expect(actualGeo2).toBeDefined();


			expect(actualGeo2.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo2.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta2 = knownHouseGeo.milesTo(actualGeo2);

			// eslint-disable-next-line no-console
			console.log('actualGeoDelta ' + actualGeoDelta2);
			expect(actualGeoDelta2).toBeLessThan(0.1);

			const spoofDelta2 = knownHouseGeo.milesTo(spoof2);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta2).toBeLessThan(1.1);
		});

		it('should geocode an 2753', async () => {
			expect.assertions(21);
			const street1 = '2753 E Windrose Dr';
			const city1 = 'Phoenix';
			const state1 = 'Arizona';
			const country1 = 'USA';
			const zipcode1 = '85032';

			// Attempt to trick the system into thinking
			// that it is a different address
			const street2 = '2753 east WindRose drive';
			const city2 = 'Phoenix';
			const state2 = 'az';
			const country2 = 'United states';
			const zipcode2 = '85032';

			const knownLat = 33.602524;
			const knownLong = -112.022703;
			const knownHouseGeo = PCAddressPreview.GeoPoint(knownLat, knownLong);
			const closeEnoughLat = 33.602;
			const closeEnoughLong = -112.022;

			const user = {
				__type: 'Pointer',
				className: '_User',
				objectId: 'abc123',
			};

			// check that caches results are the same
			const address = new PCAddressPreview();

			address.street(street1);
			address.city(city1);
			address.state(state1);
			address.country(country1);
			address.zipcode(zipcode1);
			address.user(user);
			// address.previewRadius(1); default

			const result = await address.savePreview();

			// eslint-disable-next-line no-console
			console.log('ncioed' + JSON.stringify(result));
			expect(result).toBeDefined();
			expect(result.className).toBe('PCAddressPreview');
			expect(result.id).toHaveLength(10);
			const spoofId = result.id;
			const spoof = result.get('geoPoint');

			expect(spoof).toBeDefined();

			const pcAddress = result.get('address');

			expect(pcAddress).toBeDefined();

			const actualGeo = pcAddress.get('geoPoint');

			expect(actualGeo).toBeDefined();


			expect(actualGeo.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta = knownHouseGeo.milesTo(actualGeo);

			expect(actualGeoDelta).toBeLessThan(0.1);

			const spoofDelta = knownHouseGeo.milesTo(spoof);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta).toBeLessThan(1.1);

			// check that caches results are the same
			const address2 = new PCAddressPreview();

			address2.street(street2);
			address2.city(city2);
			address2.state(state2);
			address2.country(country2);
			address2.zipcode(zipcode2);
			address2.user(user);
			// address2.previewRadius(1); default

			const result2 = await address2.savePreview();

			// eslint-disable-next-line no-console
			console.log('dif' + JSON.stringify(result2));
			expect(result2).toBeDefined();
			expect(result2.className).toBe('PCAddressPreview');
			expect(result2.id).toHaveLength(10);
			// ensures the cache was used
			expect(result2.id).toBe(spoofId);
			const spoof2 = result2.get('geoPoint');

			expect(spoof2).toBeDefined();

			const pcAddress2 = result2.get('address');

			expect(pcAddress2).toBeDefined();

			const actualGeo2 = pcAddress2.get('geoPoint');

			expect(actualGeo2).toBeDefined();


			expect(actualGeo2.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo2.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta2 = knownHouseGeo.milesTo(actualGeo2);

			// eslint-disable-next-line no-console
			console.log('actualGeoDelta ' + actualGeoDelta2);
			expect(actualGeoDelta2).toBeLessThan(0.1);

			const spoofDelta2 = knownHouseGeo.milesTo(spoof2);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta2).toBeLessThan(1.1);
		});

		it('should NOT geocode an 2707', async () => {
			expect.assertions(21);
			const street1 = '2707 E Windrose Dr';
			const city1 = 'Phoenix';
			const state1 = 'Arizona';
			const country1 = 'USA';
			const zipcode1 = '85032';

			// Attempt to trick the system into thinking
			// that it is a different address
			const street2 = '2707 east WindRose drive';
			const city2 = 'Phoenix';
			const state2 = 'az';
			const country2 = 'United states';
			const zipcode2 = '85032';

			const knownLat = 33.602525;
			const knownLong = -112.023747;
			const knownHouseGeo = PCAddressPreview.GeoPoint(knownLat, knownLong);
			const closeEnoughLat = 33.602;
			const closeEnoughLong = -112.023;

			const user = {
				__type: 'Pointer',
				className: '_User',
				objectId: 'abc123',
			};

			// check that caches results are the same
			const address = new PCAddressPreview();

			address.street(street1);
			address.city(city1);
			address.state(state1);
			address.country(country1);
			address.zipcode(zipcode1);
			address.user(user);
			// address.previewRadius(1); default

			const result = await address.savePreview();

			// eslint-disable-next-line no-console
			console.log('ncioed' + JSON.stringify(result));
			expect(result).toBeDefined();
			expect(result.className).toBe('PCAddressPreview');
			expect(result.id).toHaveLength(10);
			const spoofId = result.id;
			const spoof = result.get('geoPoint');

			expect(spoof).toBeDefined();

			const pcAddress = result.get('address');

			expect(pcAddress).toBeDefined();

			const actualGeo = pcAddress.get('geoPoint');

			expect(actualGeo).toBeDefined();


			expect(actualGeo.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta = knownHouseGeo.milesTo(actualGeo);

			expect(actualGeoDelta).toBeLessThan(0.1);

			const spoofDelta = knownHouseGeo.milesTo(spoof);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta).toBeLessThan(1.1);

			// check that caches results are the same
			const address2 = new PCAddressPreview();

			address2.street(street2);
			address2.city(city2);
			address2.state(state2);
			address2.country(country2);
			address2.zipcode(zipcode2);
			address2.user(user);
			// address2.previewRadius(1); default

			const result2 = await address2.savePreview();

			// eslint-disable-next-line no-console
			console.log('dif' + JSON.stringify(result2));
			expect(result2).toBeDefined();
			expect(result2.className).toBe('PCAddressPreview');
			expect(result2.id).toHaveLength(10);
			// ensures the cache was used
			expect(result2.id).toBe(spoofId);
			const spoof2 = result2.get('geoPoint');

			expect(spoof2).toBeDefined();

			const pcAddress2 = result2.get('address');

			expect(pcAddress2).toBeDefined();

			const actualGeo2 = pcAddress2.get('geoPoint');

			expect(actualGeo2).toBeDefined();


			expect(actualGeo2.latitude + '').toContain(closeEnoughLat + '');
			expect(actualGeo2.longitude + '').toContain(closeEnoughLong + '');


			// the actual geo should be within 0.1 mile of our known geo
			// give 1/10 of a mile flex room
			const actualGeoDelta2 = knownHouseGeo.milesTo(actualGeo2);

			// eslint-disable-next-line no-console
			console.log('actualGeoDelta ' + actualGeoDelta2);
			expect(actualGeoDelta2).toBeLessThan(0.1);

			const spoofDelta2 = knownHouseGeo.milesTo(spoof2);

			// the spoof geo should be within 1 mile of our known geo
			expect(spoofDelta2).toBeLessThan(1.1);
		});

		it('check totals', async () => {
			expect.assertions(4);

			const query1 = PCAddressPreview.Query('PCAddress');

			const result1 = await query1.find();

			expect(result1).toBeDefined();
			expect(result1).toHaveLength(3);

			const query2 = PCAddressPreview.Query('PCAddressPreview');

			const result2 = await query2.find();

			expect(result2).toBeDefined();
			expect(result2).toHaveLength(3);
		});
	});
});



src/PCAddressPreview.js

const TestingParse = require('parse/node');

const PCGeocoder = require('@panda-clouds/geocoder');
const PCAddressFormatter = require('@panda-clouds/address-formatter');

const defaultRadius = 1;


class PCAddressPreview {
	constructor() {
		// Empty Constructor
		// default to 1 mile
		this.radius = defaultRadius;
	}
	static Object(className) {
		if (typeof Parse !== 'undefined') {
			/* global Parse */
			return new Parse.Object(className);
		}

		return new TestingParse.Object(className);
	}
	static Query(className) {
		if (typeof Parse !== 'undefined') {
			/* global Parse */
			return new Parse.Query(className);
		}

		return new TestingParse.Query(className);
	}
	static GeoPoint(lat, long) {
		if (typeof Parse !== 'undefined') {
			/* global Parse */
			return new Parse.GeoPoint(lat, long);
		}

		return new TestingParse.GeoPoint(lat, long);
	}
	static bestParse() {
		if (typeof Parse !== 'undefined') {
			/* global Parse */
			return Parse;
		}

		return TestingParse;
	}
	static _generateRandomFloat(min, max, decimals) {
		// Truly random
		// https://gist.github.com/naomik/6030653
		return parseFloat((Math.random() * (max - min) + min).toFixed(decimals));
	}

	static _randomPointWithInRadiusInMiles(centerGeo, radiusInput) {
		// Lat is about 69 miles apart always
		// Long are about 69 miles apart at the equator
		// Long in 0 miles at the north pole
		// https://www.thoughtco.com/degree-of-latitude-and-longitude-distance-4070616
		let radius;

		if (radiusInput) {
			radius = radiusInput;
		} else {
			radius = defaultRadius;
		}

		const aboutOneMileInDegrees = 1 / 69;
		const radiusInDegrees = aboutOneMileInDegrees * radius;

		let randomGeo;
		let distance;
		// Why a do-while?
		// the 1/69 is an esimate so we check that
		// the dot actually falls within the radius

		do {
			const latDelta = PCAddressPreview._generateRandomFloat(-radiusInDegrees, radiusInDegrees, 6);
			const longDelta = PCAddressPreview._generateRandomFloat(-radiusInDegrees, radiusInDegrees, 6);

			const randLat = centerGeo.latitude + latDelta;
			const randLong = centerGeo.longitude + longDelta;

			randomGeo = PCAddressPreview.GeoPoint(randLat, randLong);

			distance = centerGeo.milesTo(randomGeo);
		} while (distance > radius || distance < 0.02);

		// Make sure the point in within the X mile radius
		// and not within 0.02 miles (105.6 feet) of the house.
		// (a spoof is usless if its right on the house)
		return randomGeo;
	}
	async _save(result) {
		// result comes from PCGeocoder
		// its an object with:
		// - lat
		// - long
		// - address
		// -- {'formattedAddress':'2753 E Windrose Dr, Phoenix, AZ 85032, United States','latitude':33.6025,'longitude':-112.02269,'country':'United States','countryCode':'US','state':'Arizona','county':'Maricopa','city':'Phoenix','zipcode':'85032','district':'North Phoenix','streetName':'E Windrose Dr','streetNumber':'2753','building':null,'extra':{'herePlaceId':'NT_K9IhbsLP54qz5bbKFLPjcA_ycTNzA','confidence':1},'administrativeLevels':{'level1long':'Arizona','level2long':'Maricopa'},'provider':'here'}
		// -raw (Depends on the geocoder that delivered the results)
		// -- [{'place_id':'203032233','licence':'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright','osm_type':'way','osm_id':'5609650','boundingbox':['33.602592','33.602692','-112.022642','-112.022542'],'lat':'33.602642','lon':'-112.022592','display_name':'2753, East Windrose Drive, Phoenix, Maricopa County, Arizona, 85032, USA','class':'place','type':'house','importance':0.7409999999999999,'address':{'house_number':'2753','road':'East Windrose Drive','city':'Phoenix','county':'Maricopa County','state':'Arizona','postcode':'85032','country':'USA','country_code':'us'}}

		if (!result.address) {
			return PCAddressPreview.bestParse().Promise.reject('Missing Address Object');
		}

		let shortState;

		if (result.address.state) {
			shortState = PCAddressFormatter.state(result.address.state);
		}

		const cache = PCAddressPreview.Object('PCAddress'); // convert 'Arizona' to 'AZ'

		if (this.user) {
			cache.set('creator', this.user);
		}

		const geo = PCAddressPreview.GeoPoint(result.address.latitude, result.address.longitude);

		if (this.nickname) {
			cache.set('name', this.nickname);
		}

		if (this.notes) {
			cache.set('notes', this.notes);
		}

		if (this.gateCode) {
			cache.set('gateCode', this.gateCode);
		}

		if (this.unit) {
			cache.set('unit', this.unit);
		}

		if (geo) {
			cache.set('geoPoint', geo); // (33.6025,-112.02269)
		}

		cache.set('streetNumber', result.address.streetNumber); // '2753'
		cache.set('streetName', result.address.streetName);
		cache.set('street', result.address.street); // 'E Windrose Dr'

		if (this.street !== result.address.street) {
			cache.set('inputStreet', this.street);
		}

		cache.set('city', result.address.city); // 'Phoenix'

		if (this.city !== result.address.city) {
			cache.set('inputCity', this.city);
		}

		cache.set('county', result.address.county); // 'Maricopa'
		cache.set('state', shortState); // 'AZ'

		if (this.state !== shortState) {
			cache.set('inputState', this.state);
		}

		// cache.set('country',result.address.country); X  // 'United States' We decided to use country code for less storage
		cache.set('country', result.address.countryCode); // 'US'

		if (this.country !== result.address.countryCode) {
			cache.set('inputCountry', this.country);
		}

		const zipOnly = PCAddressFormatter.zipcode(result.address.zipcode);

		if (!this.plusFour) {
			// the user didn't input it,
			// lets see if the provider returned it?
			// mapquest and sometimes openstreetmap return the +4
			this.plusFour = PCAddressFormatter.plusFour(result.address.zipcode);
		}

		cache.set('zipcode', zipOnly); // '85032'

		if (this.zipcode !== zipOnly) {
			cache.set('inputZipcode', this.zipcode);
		}

		if (this.plusFour) {
			cache.set('plusFour', this.plusFour);
		}

		cache.set('provider', result.address.provider); // 'here'

		// cache.set('formattedAddress',result.address.formattedAddress)    X   // '2753 E Windrose Dr, Phoenix, AZ 85032, United States' We decided to force manual formatting for less storage
		const address = await cache.save(null, this.permissions());


		const spoof = PCAddressPreview.Object('PCAddressPreview');
		const oneMileSpoof = PCAddressPreview._randomPointWithInRadiusInMiles(geo, this.radius);

		spoof.set('radiusInMiles', this.radius);
		spoof.set('geoPoint', oneMileSpoof);
		spoof.set('address', address.toPointer());
		spoof.set('city', result.address.city);
		spoof.set('zipcode', zipOnly);

		if (this.user) {
			spoof.set('creator', this.user);
		}

		const final = await spoof.save(null, this.permissions());

		return final;
	}

	async _searchCache() {
		const innerQuery = PCAddressPreview.Query('PCAddress');

		innerQuery.equalTo('street', this.street);
		innerQuery.equalTo('city', this.city);
		innerQuery.equalTo('state', this.state);
		innerQuery.equalTo('country', this.country);
		innerQuery.equalTo('zipcode', this.zipcode);
		innerQuery.equalTo('creator', this.user);

		const query = PCAddressPreview.Query('PCAddressPreview');

		query.matchesQuery('address', innerQuery);
		query.equalTo('radiusInMiles', this.radius);
		query.include('address');

		const first = await query.first(this.permissions());

		return first;
	}

	nickname(input) {
		this.nickname = input;
	}

	notes(input) {
		this.notes = input;
	}

	gateCode(input) {
		this.gateCode = input;
	}

	plusFour(input) {
		this.plusFour = input;
	}

	unit(input) {
		this.unit = input;
	}

	street(input) {
		this.street = PCAddressFormatter.street(input);
		const maybeUnit = PCAddressFormatter.unit(input);

		if (maybeUnit) {
			this.unit = maybeUnit;
		}
	}

	city(input) {
		this.city = PCAddressFormatter.city(input);
	}

	state(input) {
		this.state = PCAddressFormatter.state(input);
	}

	country(input) {
		this.country = PCAddressFormatter.country(input);
	}

	zipcode(input) {
		this.zipcode = PCAddressFormatter.zipcode(input);
		const maybePlusFour = PCAddressFormatter.plusFour(input);

		if (maybePlusFour) {
			this.plusFour = maybePlusFour;
		}
	}

	// addr.user(request.user)
	user(input) {
		this.user = input;
	}

	useMasterKey(input) {
		this.useMasterKeyBoolean = input;
	}

	permissions() {
		const permDic = {};

		if (this.useMasterKeyBoolean) {
			permDic.useMasterKey = this.useMasterKeyBoolean;
		}

		if (this.user && this.user.getSessionToken) {
			permDic.sessionToken = this.user.getSessionToken();
		}

		return permDic;
	}

	previewRadius(input) {
		if (isNaN(input)) {
			throw new Error('previewRadius expects a number');
		}

		if (input > 99999) {
			throw new Error('Please specifiy a number less than 999 miles');
		}

		if (input < 0.05) {
			throw new Error('Please specifiy a number greater than 0.05 miles (264\' is only 4-6 houses)');
		}

		this.radius = input;
	}

	// primaryProviders are attemped before 'free services'
	primaryProviders(input) {
		// geo.primaryProviders( [{provider:'locationiq',apiKey:'abc123'},{provider:'google',apiKey:'def345'}] )
		this.primaryProvidersArray = input;
	}

	// backupProviders are attemped after 'free services'
	backupProviders(input) {
		// geo.backupProviders( [{provider:'locationiq',apiKey:'abc123'},{provider:'google',apiKey:'def345'}] )
		this.backupProvidersArray = input;
	}

	async savePreview() {
		// This function first tries to query our database to:
		// 1. reduce the dependency on external geocoder API's
		// 2. to keep a consistent randomMarker (to keep malicous users
		//    from refreshing the random markers and ruling out areas)
		// 3. to speed up response time
		// 4. to keep perfectly consistent coordinates

		const result = await this._searchCache();

		if (result) {
			// We have cached results
			return result;
		}

		// Not found.
		// Geocode then cache
		const geo = new PCGeocoder();

		geo.street(this.street);
		geo.city(this.city);
		geo.state(this.state);
		geo.country(this.country);
		geo.zipcode(this.zipcode);

		// forward provider info
		if (this.primaryProvidersArray) {
			geo.primaryProviders(this.primaryProvidersArray);
		}

		if (this.backupProvidersArray) {
			geo.backupProviders(this.backupProvidersArray);
		}

		const searchResults = await geo.search();

		if (searchResults) {
			await this._save(searchResults);
			const maybeNewResults = await this._searchCache();

			if (maybeNewResults) {
				return maybeNewResults;
			}

			throw new Error('We couldn\'t find that address');
		} else {
			throw new Error('We couldn\'t find that address');
		}
	}
}
module.exports = PCAddressPreview;
